### DISCUSSION TITLE
ArrayIndexOutOfBoundsException

### DISCUSSION THREAD

**Discussion by Lineflyer** [mapsforge/mapsforge#1233] [Q&A] [answered]
We are using the mapsforge library in our geocaching open source app [c:geo](https://github.com/cgeo/cgeo).

Since some months we do see a crash report on Google Play with the following typical stacktrace:
```
java.lang.ArrayIndexOutOfBoundsException: 
  at java.util.concurrent.CopyOnWriteArrayList.get (CopyOnWriteArrayList.java:380)
  at java.util.concurrent.CopyOnWriteArrayList.get (CopyOnWriteArrayList.java:393)
  at org.mapsforge.map.layer.Layers.get (Layers.java:336)
  at org.mapsforge.map.android.input.TouchGestureHandler.onSingleTapConfirmed (TouchGestureHandler.java:263)
  at android.view.GestureDetector$GestureHandler.handleMessage (GestureDetector.java:323)
  at android.os.Handler.dispatchMessage (Handler.java:107)
  at android.os.Looper.loop (Looper.java:214)
  at android.app.ActivityThread.main (ActivityThread.java:7386)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:492)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:980)
```

We could not yet forcefully trigger or reproduce this issue, however a user once reported it happens after closing and opening the map view a few times. (Refer to: https://github.com/cgeo/cgeo/issues/10128)

I wonder whether you have any idea (in case the problem is caused by our code) or could check whether there is a possibility to fix it in the mapsforge code.

We are currently using version 0.16.0 with the previous version 0.15.0.

**Comment by Lineflyer** [comment]
_Originally posted by @eddiemuc in https://github.com/cgeo/cgeo/issues/10128#issuecomment-797688798_

The crash above looks like a problem in mapsforge. Maybe indexes of layers are not tracked in a thread-safe way and thus many additions/deletions of layers might bring it out of sync (thus the indexoutofbounds).

**Comment by devemux86** [comment]
I have not seen such exception and can not know your implementation, how you use threads in your app
or if you destroy / (re)create constantly the map view (not [recommended](https://github.com/mapsforge/mapsforge/blob/master/mapsforge-samples-android/AndroidManifest.xml#L69) workflow) on device rotation, etc.

In any case the `Layers` calls appear `synchronized`.

If you are recently seeing this exception, you may need to debug the recent changes in your app code.

**Comment by eddiemuc** [comment]
Hi @devemux86 , sorry to bother you with this again, but we keep having problems with above exception and are a bit lost what to do.

I checked the code in org.mapsforge.map.android.input.TouchGestureHandler.onSingleTapConfirmed, and I believe there's a chance that it produces an ArrayIndexOutOfBound exception. This will happen if the layer list is reduces/deleted between code line 263 (where size is queried) and the loop in 264 (where list is iterated, assuming that size will not change during iteration).

If an app e.g. clears the layer list while the loop is done, this will lead to ArrayIndexOutOfBound.

Question: would you in principal accept a pull request from me to fix this problem (after review of course)? Then I would start working on such a PR.

Thanks and best regards
  Eddie

**Comment by devemux86** [comment]
@eddiemuc thanks for the check.

What change do you have in mind?
Have you confirmed that it really solves the problem in your app?

To avoid experiments with upstream, you could include in your app a modified Mapsforge
and if it solves your issue, then we can discuss the change which has proven successful.

**Comment by eddiemuc** [comment]
@devemux86 

The issue is hard to reproduce since it involves race conditions. So it is hard to test this in a development environment. We see this issue frequently on Google Console however.

The change I have in mind currently is as follows:
* In class Layers: introduce a method "getIfValid(int index)". in contrast to the "get(int index)" method it checks the index for validity (>=0 and < size()), then returning either the requested element or null in case of invalid index. getIfValid would be synchronized, so nothing bad can happen here.
*  In class TouchGestureHandler: in the affected loop, use the new getIfValid-method instead of "get". If null is returned then the entry is skipped (or maybe the loop can be aborted as a whole)

I think this would be a light-weighted, low-risk change to get the wanted behaviour/stability. What do you think?

**Reply by devemux86** [reply]
As this is only reported for your app and everyone else has not seen such problem:
You can change Mapsforge code, include it in your app, release it and check if it solves your case.
(publish Mapsforge in local Maven with `publishToMavenLocal` and include it in your app)

Regarding your suggestion, if only to check if `i` is within list limits, we can check in the loop if the `0 ≤ i ≤ layers.size() - 1`?

**Reply by eddiemuc** [reply]
> Regarding your suggestion, if only to check if `i` is within list limits, we can check in the loop if the `0 ≤ i ≤ layers.size() - 1`?

In that case the check would not be synchronized. Between the call to "size()" and the call to "get()" the "size()" might be changed by another thread. However, if you don't want to change the "Layers", this can be also synchronized in TouchGestureHandler using a separate synchronized e.g. as follows:

```
synchronized(this.mapView.getLayerManager().getLayers()) {
   if (i >= this.mapView.getLayerManager().getLayers().size()) {
       continue;
   }
   Layer layer = this.mapView.getLayerManager().getLayers().get(i)
}
```

(get() and size() must be in same synchronized block)

This would work too. Would you prefer a pattern like this?

**Reply by devemux86** [reply]
Then the whole `for` could be inside the `synchronized` block?

And instead of `continue`, a `break` could be needed, as it indicates that the loop became invalid.

**Reply by eddiemuc** [reply]
> Then the whole `for` could be inside the `synchronized` block?

I wouldn't do that. While this works synchronization-wise, it will make the thread hold the mutex for a long time (esp when the list of layers is long) which might lead to blocking issues in other parts of the app. After all, the indexoutofbounds is a seldom case and shouldn't lead to such frequent blockers IMHO.
 
> And instead of `continue`, a `break` could be needed, as it indicates that the loop became invalid.

Sure, sorry, the above code was just to show the pattern...

**Reply by devemux86** [reply]
Then let's try your suggestion, please create a pull request to see the code changes.

But before merging, I suggest to include them in your app and see first if they solve your report.

**Comment by eddiemuc** [comment]
I just saw that in TouchGestureHandler, the size-loop-pattern is used in multiple places. So I guess it would make sense to introduce the above pattern to all of these places

**Comment by eddiemuc** [comment]
> As this is only reported for your app and everyone else has not seen such problem:

The cause for this might be that we have often very large lists of layers (several hundert cache icons displayed at once) which are then cleared in a single command (as offered by the Layers class). Apart from that, I think the current code has a vulnerability here - even if it is not visible in many apps

**Comment by eddiemuc** [comment, marked as answer]
Hi @devemux86 

I created PR #1256 

But I have problems getting mapsforge running on my local machine: compile step does not find any of the android-specific classes.

I see that there are recent changes in the gradle files with upgrade to Gradle 7 (although in grade, 4.4.2 still seems to be configured), There seems to be a need for some environment variable ANDROID_HOME now pointing to some specific directory substructure.  Do you have any instruction what I need to do to get mapsforge android parts compiled locally?

Thanks and regards
  Eddie

**Reply by devemux86** [reply]
Gradle build system vs Android Gradle plugin are not necessarily connected.
Android Gradle plugin 4.2.2 is used because IDEA is still not compatible with Android Gradle plugin 7.0.x.

ANDROID_HOME should be set to the installed Android SDK location, e.g. "/home/profile/Android/Sdk"
GitHub actions are configured that way and they build the repository successfully.
See also the Android documentation: https://developer.android.com/studio/command-line/variables

**Reply by eddiemuc** [reply]
Thanks a lot, works like a charm now
